Source/CarthageKit/Project.swift

```diff
@@ -587,21 +587,81 @@ public final class Project { // swiftlint:disable:this type_body_length
 	) -> SignalProducer<URL, CarthageError> {
+		
+		typealias FrameworkUrlAndDestinationUrl = (frameworkUrlInUnarchivedFolder: URL, frameworkUrlInBuildForlder: URL)
+
 		return SignalProducer<URL, CarthageError>(value: zipFile)
 			.flatMap(.concat, unarchive(archive:))
 			.flatMap(.concat) { directoryURL -> SignalProducer<URL, CarthageError> in
+				// For all frameworks in the directory where the archive has been expanded
 				return frameworksInDirectory(directoryURL)
-					.flatMap(.merge) { url -> SignalProducer<URL, CarthageError> in
-						return checkFrameworkCompatibility(url, usingToolchain: toolchain)
+					.collect()
+					// Check if multiple framewroks resolve to the same unique destination URL in the Carhage/Build/ folder.
+					// This is needed because frameworks might overwrite each others.
+					.flatMap(.merge) { frameworksUrls -> SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError> in
+						return SignalProducer<URL, CarthageError>(frameworksUrls)
+							.flatMap(.merge) { url -> SignalProducer<URL, CarthageError> in
+								return platformForFramework(url)
+									.attemptMap { platform -> Result<URL, CarthageError> in
+										guard let destinationURLInWorkingDir = platform
+											.relativeURL?
+											.appendingPathComponent(url.lastPathComponent, isDirectory: true) else {
+												return .failure(.internalError(description: "failed to construct framework destination url from \(platform) and \(url.lastPathComponent)"))
+										}
+
+										return .success(self
+											.directoryURL
+											.appendingPathComponent(destinationURLInWorkingDir.path, isDirectory: true))
+								}
+							}
+							.collect()
+							.flatMap(.merge) { destinationUrls -> SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError> in
+
+								let frameworkUrlAndDestinationUrlPairs = zip(frameworksUrls.map{$0.standardizedFileURL}, destinationUrls.map{$0.standardizedFileURL})
+									.map { FrameworkUrlAndDestinationUrl(frameworkUrlInUnarchivedFolder:$0, frameworkUrlInBuildForlder: $1)}
+
+								let destinationMap = frameworkUrlAndDestinationUrlPairs.reduce([URL: [URL]]()) { dict, pair in
+									var mutableDict = dict
+									mutableDict[pair.frameworkUrlInBuildForlder] = (mutableDict[pair.frameworkUrlInBuildForlder] ?? []) + [pair.frameworkUrlInUnarchivedFolder]
+
+									return mutableDict
+								}
+
+								let dupes = destinationMap.filter { $0.value.count > 1 }
+
+								guard dupes.count == 0 else {
+
+									return SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError>(error: CarthageError.duplicatesInArchive(duplicates: dupes))
+								}
+
+								let uniquePairs = destinationMap
+									.filter { $0.value.count == 1}
+									.map { FrameworkUrlAndDestinationUrl(frameworkUrlInUnarchivedFolder:$0.key,
+																		 frameworkUrlInBuildForlder: $0.value.first!)}
+
+								return SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError>(uniquePairs)
```

Comment:
Same here. Could this be extracted to a helper function?

I find the code much easier to follow when the `flatMap`s are very high level.

Reply:
I only moved it out of the flatMap but it's still inside the function. The purpose of the function is really to `uniquePairs` keyed by the first element of the pair.  I didn't make it super generic because I still kept the context, but I could. However I need a `GenericError` since result is not Either and only takes `: Error`.

Something like

```
		func uniquePairs<A, B>(
			_ pairs: [(A, B)]
			) -> Result<[(A, B>], GenericError<[A: [B]]>> {
			let map = pairs
				.reduce([A: [B]]()) { dict, pair in
					var mutableDict = dict
					mutableDict[pair.0] =
						(mutableDict[pair.0] ?? []) + [pair.1]

					return mutableDict
			}

			let dupes = map.filter { $0.value.count > 1 }
			guard dupes.count == 0 else {
				return .failure(GenericError<[A: [B]]>(value: dupes))
			}

			let uniquePairs = map
				.filter { $0.value.count == 1}
				.map { ($0, $0.value.first!)}
			return .success(uniquePairs)
		}
```